clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/server/P2PFileTransferServerHandler.java

```diff
@@ -92,90 +104,121 @@ protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest httpReque
+      snapshotDir = new File(blobTransferRequest.getSnapshotDir());
 
-        // Check the snapshot table format
-        BlobTransferTableFormat currentSnapshotTableFormat = blobSnapshotManager.getBlobTransferTableFormat();
-        if (blobTransferRequest.getRequestTableFormat() != currentSnapshotTableFormat) {
-          byte[] errBody = ("Table format mismatch for " + blobTransferRequest.getFullResourceName()
-              + ", current snapshot format is " + currentSnapshotTableFormat.name() + ", requested format is "
-              + blobTransferRequest.getRequestTableFormat().name()).getBytes();
-          setupResponseAndFlush(HttpResponseStatus.NOT_FOUND, errBody, false, ctx);
-          return;
-        }
+      // Check the snapshot table format
+      BlobTransferTableFormat currentSnapshotTableFormat = blobSnapshotManager.getBlobTransferTableFormat();
+      if (blobTransferRequest.getRequestTableFormat() != currentSnapshotTableFormat) {
+        byte[] errBody = ("Table format mismatch for " + blobTransferRequest.getFullResourceName()
+            + ", current snapshot format is " + currentSnapshotTableFormat.name() + ", requested format is "
+            + blobTransferRequest.getRequestTableFormat().name()).getBytes();
+        setupResponseAndFlush(HttpResponseStatus.NOT_FOUND, errBody, false, ctx);
+        return;
+      }
 
-        try {
-          transferPartitionMetadata =
-              blobSnapshotManager.getTransferMetadata(blobTransferRequest, successCountedAsActiveCurrentUser);
-        } catch (Exception e) {
-          setupResponseAndFlush(HttpResponseStatus.NOT_FOUND, e.getMessage().getBytes(), false, ctx);
-          return;
-        }
+      // Check the concurrent request limit
+      if (globalConcurrentTransferRequests.get() >= maxAllowedConcurrentSnapshotUsers) {
+        String errMessage =
+            "The number of concurrent snapshot users exceeds the limit of " + maxAllowedConcurrentSnapshotUsers
+                + ", wont be able to process the request for " + blobTransferRequest.getFullResourceName();
+        LOGGER.error(errMessage);
+        setupResponseAndFlush(HttpResponseStatus.TOO_MANY_REQUESTS, errMessage.getBytes(), false, ctx);
+        return;
+      }
 
-        if (!snapshotDir.exists() || !snapshotDir.isDirectory()) {
-          byte[] errBody = ("Snapshot for " + blobTransferRequest.getFullResourceName() + " doesn't exist").getBytes();
-          setupResponseAndFlush(HttpResponseStatus.NOT_FOUND, errBody, false, ctx);
-          return;
+      try {
+        transferPartitionMetadata =
+            blobSnapshotManager.getTransferMetadata(blobTransferRequest, successCountedAsActiveCurrentUser);
+        ctx.channel().attr(SUCCESS_COUNTED).set(successCountedAsActiveCurrentUser);
+        ctx.channel().attr(BLOB_TRANSFER_REQUEST).set(blobTransferRequest);
+        if (successCountedAsActiveCurrentUser.get()) {
+          globalConcurrentTransferRequests.incrementAndGet();
```

Comment:
We should do a check here again to avoid the race condition:
```
if (globalConcurrentTransferRequests.incrementAndGet() > maxAllowedConcurrentSnapshotUsers) {
   // throw Exception;
}
```